src/main/java/apoc/meta/Tables4LabelsProfile.java

```diff
@@ -0,0 +1,256 @@
+        if (relMap.containsKey(relType)) {
+            return relMap.get(relType);
+        } else {
+            PropertyContainerProfile p = new PropertyContainerProfile();
+            relMap.put(relType, p);
+            return p;
+        }
+    }
+
+    public Long sawNode(OrderedLabels labels) {
+        if (obsByNode.containsKey(labels)) {
+            Long val = obsByNode.get(labels) + 1;
+            obsByNode.put(labels, val);
+            return val;
+        } else {
+            obsByNode.put(labels, 1L);
+            return 1L;
+        }
+    }
+
+    public Long sawRel(String typeName) {
+        if (obsByRelType.containsKey(typeName)) {
+            Long val = obsByRelType.get(typeName) + 1;
+            obsByRelType.put(typeName, val);
+            return val;
+        } else {
+            obsByRelType.put(typeName, 1L);
+            return 1L;
+        }
+    }
+
+    public void observe(Node n, MetaConfig config, Iterable<ConstraintDefinition> constraints, Map<String, Iterable<ConstraintDefinition>> relConstraints) {
+        OrderedLabels labels = new OrderedLabels(n.getLabels());
+        PropertyContainerProfile localNodeProfile = getNodeProfile(labels);
+
+        Set<String> excludes = config.getExcludes();
+        Set<String> includesRels = config.getIncludesRels();
+
+        // Only descend and look at properties if it's in our match list.
+        if (config.matches(n.getLabels())) {
+            sawNode(labels);
+            localNodeProfile.observe(n, constraints, true, relConstraints, null, null);
+        }
+
+        // Even if the node isn't in our match list, do rel processing.  This
+        // is because our profiling is "node-first" to get to the relationships,
+        // and if we don't do it this way, it's possible to blacklist nodes and
+        // thereby miss relationships that were of interest.
+        for (RelationshipType type : n.getRelationshipTypes()) {
+            String typeName = type.name();
+
+            if (!config.matches(type)) { continue; }
+
+            int out = n.getDegree(type, Direction.OUTGOING);
+            if (out == 0) continue;
+
+            long seenSoFar = sawRel(typeName);
+
+            if (seenSoFar > config.getMaxRels()) {
+                // We've seen more than the maximum sample size for this rel, so
+                // we don't need to keep looking.
+                continue;
+            }
+
+            PropertyContainerProfile localRelProfile = getRelProfile(typeName);
+
+            for(Relationship r : n.getRelationships(type, Direction.OUTGOING)) {
+                Iterable<Label> relStartNode = r.getStartNode().getLabels();
+                Iterable<Label> relEndNode = r.getEndNode().getLabels();
+                boolean isNode = false;
+                localRelProfile.observe(r, constraints, isNode, relConstraints, relStartNode, relEndNode);
+            }
+        }
+    }
+
+    public Tables4LabelsProfile finished() {
```

Comment:
It's here, when all processing is completed -- that you should consider the constraints and do it only once.   Don't observe the constraints inside of the 'observe' method that will be called potentially millions of times.  It'll also help refactor that deeply nested loop below where your comments say there's a performance concern.

Reply:
Moved the check code to finished(), and cleaned up a lot of the passing around of constraint data.